core/blockchain.go

```diff
@@ -231,10 +229,10 @@ func (bc *BlockChain) SetHead(head uint64) {
 
 	// Update all computed fields to the new head
 	if bc.currentBlock != nil && bc.hc.CurrentHeader().Number.Uint64() < bc.currentBlock.NumberU64() {
-		bc.currentBlock = bc.GetBlock(bc.hc.CurrentHeader().Hash())
+		bc.currentBlock = bc.GetBlock(bc.hc.CurrentHeader().Hash(), bc.hc.CurrentHeader().Number.Uint64())
```

Comment:
Although we can query the current header here multiple times consecutively safely because the blockchain's mutex is locked, it looks and feels wrong and a code refactor could break it. Please fetch the current header once and access it fields for both the hash and number.